patches/net/minecraft/core/MappedRegistry.java.patch

```diff
@@ -26,6 +41,41 @@
        }
  
        this.byKey.put(p_256594_, reference);
+@@ -162,6 +_,7 @@
+       this.lifecycles.put(p_256374_, p_256469_);
+       this.registryLifecycle = this.registryLifecycle.add(p_256469_);
+       this.holdersInOrder = null;
++      this.addCallbacks.forEach(addCallback -> addCallback.onAdd(this, p_256563_, p_256594_, p_256374_));
+       return reference;
+    }
+ 
+@@ -190,7 +_,7 @@
+    @Nullable
+    @Override
+    public T get(@Nullable ResourceKey<T> p_122714_) {
+-      return getValueFromNullable(this.byKey.get(p_122714_));
++      return getValueFromNullable(this.byKey.get(resolve(p_122714_)));
+    }
+ 
+    @Nullable
+@@ -206,7 +_,7 @@
+ 
+    @Override
+    public Optional<Holder.Reference<T>> getHolder(ResourceKey<T> p_205905_) {
+-      return Optional.ofNullable(this.byKey.get(p_205905_));
++      return Optional.ofNullable(this.byKey.get(resolve(p_205905_)));
+    }
+ 
+    @Override
+@@ -216,7 +_,7 @@
+    }
+ 
+    Holder.Reference<T> getOrCreateHolderOrThrow(ResourceKey<T> p_248831_) {
+-      return this.byKey.computeIfAbsent(p_248831_, p_258169_ -> {
++      return this.byKey.computeIfAbsent(resolve(p_248831_), p_258169_ -> {
```

Comment:
If I am correct that the main use of the alias system is so when things are renamed, is it possible for us to optimize this so that it returns/caches the Holder.Reference for the non aliased version? That way there is one less pointer to the alias' name making it easier data wise for mods to later remove, AND also would cut down on extra object creation. Or is it not worth doing this because the patch overhead and maintenance for it would not be worth it?

Reply:
I just pushed a fix that made `resolve` non-allocating for the common path:
![image](https://github.com/neoforged/NeoForge/assets/13494793/b78ba072-443c-4daf-9c33-199f00d4c75b)

In any case this is a `computeIfAbsent` call, so it will cache entries?

Reply:
Yes it will cache a singular extra `Holder.Reference` object, but my point was more what is the purpose of the alias system? Does it make more sense to have `getOrCreateHolderOrThrow(aliasKey).key()` return `aliasKey` or the actual main key?

Secondary question (as I have not touched it at much and following the full code path is a bit of a pain)... Does the `Holder.Reference` for the aliased entry properly have the value get (re)bound?

Reply:
Ah yes, the alias system is used for renames.

> Does it make more sense to have `getOrCreateHolderOrThrow(aliasKey).key()` return `aliasKey` or the actual main key?

It should return the main key.

> Does the `Holder.Reference` for the aliased entry properly have the value get (re)bound?

Since it will be the same holder as the resolved (i.e. main) key, it will follow the same lifecycle.

Reply:
I am dumb, for some reason overlooked the fact that resolve makes the alias return the non alias’s key.